### PR TITLE
feat: add flutter-version datasource

### LINKED ISSUE #12820: add flutter-version datasource
### What would you like Renovate to be able to do?

We need a flutter datasource to solve those issues:

- #6049
- #11132
- https://github.com/containerbase/buildpack/pull/24

### If you have any ideas on how this should be implemented, please tell us here.

should be like `adoptium-java`, https://storage.googleapis.com/flutter_infra/releases/releases_linux.json can be used to fetch available versions

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by zeshuaro**
## Changes

- Add Flutter datasource
- Issue suggested to use this link to fetch the Flutter releases https://storage.googleapis.com/flutter_infra/releases/releases_linux.json. But it's no longer updated and the new link seems to be this one instead https://storage.googleapis.com/flutter_infra_release/releases/releases_linux.json

## Context

Close #12820 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required (not sure if I need to add a README for this new datasource)

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zeshuaro** [comment]
I included the full API response for the test, not sure if we need it or a subset of it would be sufficient.

File: lib/modules/datasource/flutter/index.spec.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      httpMock
+        .scope(baseUrl)
+        .get(urlPath)
+        .reply(200, Fixtures.get('index.json'));
+      const res = await getPkgReleases({
+        datasource,
+        depName,
+      });
+      expect(res).toMatchSnapshot();
+      expect(res.releases).toHaveLength(417);
```
Comment (viceice) [inline lib/modules/datasource/flutter/index.spec.ts:51]: Please trim the JSON file to only a few versions.

**Comment by rarkins** [comment]
Is this datasource for the version of Flutter itself? For other similar ones we use a `-version` suffix to try to avoid confusion, e.g.

image

File: lib/modules/datasource/flutter/index.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  constructor() {
+    super(FlutterDatasource.id);
+  }
+
+  override readonly customRegistrySupport = false;
+
+  override readonly defaultRegistryUrls = ['https://storage.googleapis.com'];
+
+  override readonly caching = true;
```
Comment (viceice) [inline lib/modules/datasource/flutter/index.ts:16]: Isn't that the default?

File: lib/modules/datasource/flutter/index.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      releases: [],
+    };
+    try {
+      const resp = (
+        await this.http.getJson<FlutterRelease>(
+          `${registryUrl}/flutter_infra_release/releases/releases_linux.json`
+        )
+      ).body;
+      result.releases.push(
+        ...resp.releases.map(({ version, release_date, channel }) => ({
```
Comment (viceice) [inline lib/modules/datasource/flutter/index.ts:38]: Do not use push, use direct assignment instead.

File: lib/modules/datasource/flutter/types.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+export interface FlutterRelease {
+  base_url: string;
+  current_release: {
+    beta: string;
+    dev: string;
+    stable: string;
+  };
+  releases: {
```
Comment (viceice) [inline lib/modules/datasource/flutter/types.ts:8]: Please create interfaces for child types.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/flutter/index.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig, ReleaseResult } from '../types';
+import type { FlutterRelease } from './types';
+
+export class FlutterDatasource extends Datasource {
+  static readonly id = 'flutter';
```
Comment (viceice) [inline lib/modules/datasource/flutter/index.ts:6]:
```suggestion
  static readonly id = 'flutter-version';
```
Rename to `flutter-version` as @rarkins suggests

File: lib/modules/datasource/flutter/index.ts (hunk 2 above)
Comment (zeshuaro) [reply lib/modules/datasource/flutter/index.ts:16]: There's no default value set in `Datasource` ([source](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/datasource.ts#L16)), unless `undefined` means the same as setting it to `true`?

File: lib/modules/datasource/flutter-version/index.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      const resp = (
+        await this.http.getJson<FlutterResponse>(
+          `${registryUrl}/flutter_infra_release/releases/releases_linux.json`
+        )
+      ).body;
+      const stableRegex = /^\d+\.\d+\.\d+$/;
+      result.releases = resp.releases
+        // The API response contains a stable version being released as a non-stable
+        // release. And so we filter out these releases here.
+        .filter(({ version, channel }) => {
```
Comment (zeshuaro) [inline lib/modules/datasource/flutter-version/index.ts:41]: After looking into the API response from https://storage.googleapis.com/flutter_infra_release/releases/releases_linux.json in more detail, it looks like there are stable versions being released as a non-stable release. See version `2.5.0` for example, it has two releases in the response, one with the `stable` channel and one with `beta`.

So I added this filter here to only retrieve the stable release for stable versions, i.e. versions with `<major>.<minor>.<patch>`.

File: lib/modules/datasource/flutter-version/types.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+interface FlutterCurrentRelease {
```
Comment (zeshuaro) [inline lib/modules/datasource/flutter-version/types.ts:1]: Not sure if we need to export these interfaces if they're only used in this file?

File: lib/modules/datasource/flutter-version/index.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig, ReleaseResult } from '../types';
+import type { FlutterResponse } from './types';
+
+export class FlutterDatasource extends Datasource {
```
Comment (viceice) [inline lib/modules/datasource/flutter-version/index.ts:5]:
```suggestion
export class FlutterVersionDatasource extends Datasource {
```

File: lib/modules/datasource/flutter-version/types.ts (hunk 7 above)
Comment (viceice) [reply lib/modules/datasource/flutter-version/types.ts:1]: Please export, they are implicit used by typescript. They also don't harm, as they are removed at runtime.

Comment (viceice) [reply lib/modules/datasource/flutter-version/types.ts:1]:
```suggestion
export interface FlutterCurrentRelease {
```

File: lib/modules/datasource/flutter-version/types.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  beta: string;
+  dev: string;
+  stable: string;
+}
+
+interface FlutterRelease {
```
Comment (viceice) [inline lib/modules/datasource/flutter-version/types.ts:7]:
```suggestion
export interface FlutterRelease {
```

File: lib/modules/datasource/flutter-version/index.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  static readonly id = 'flutter-version';
+
+  constructor() {
+    super(FlutterDatasource.id);
```
Comment (viceice) [inline lib/modules/datasource/flutter-version/index.ts:9]:
```suggestion
    super(FlutterVersionDatasource.id);
```

File: lib/modules/datasource/flutter-version/__snapshots__/index.spec.ts.snap [hunk 11]
```diff
@@ -0,0 +1,166 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`modules/datasource/flutter/index getReleases processes real data 1`] = `
```
Comment (viceice) [inline lib/modules/datasource/flutter-version/__snapshots__/index.spec.ts.snap:3]:
```suggestion
exports[`modules/datasource/flutter-version/index getReleases processes real data 1`] = `
```

File: lib/modules/datasource/api.ts [hunk 12]
```diff
@@ -9,6 +9,7 @@ import { CondaDatasource } from './conda';
 import { CrateDatasource } from './crate';
 import { DartDatasource } from './dart';
 import { DockerDatasource } from './docker';
+import { FlutterDatasource } from './flutter-version';
```
Comment (viceice) [inline lib/modules/datasource/api.ts:12]:
```suggestion
import { FlutterVersionDatasource } from './flutter-version';
```

File: lib/modules/datasource/api.ts [hunk 13]
```diff
@@ -55,6 +56,7 @@ api.set(CondaDatasource.id, new CondaDatasource());
 api.set(CrateDatasource.id, new CrateDatasource());
 api.set(DartDatasource.id, new DartDatasource());
 api.set(DockerDatasource.id, new DockerDatasource());
+api.set(FlutterDatasource.id, new FlutterDatasource());
```
Comment (viceice) [inline lib/modules/datasource/api.ts:59]:
```suggestion
api.set(FlutterVersionDatasource.id, new FlutterVersionDatasource());
```

File: lib/modules/datasource/flutter-version/index.spec.ts [hunk 14]
```diff
@@ -0,0 +1,54 @@
+import { getPkgReleases } from '..';
+import { Fixtures } from '../../../../test/fixtures';
+import * as httpMock from '../../../../test/http-mock';
+import { EXTERNAL_HOST_ERROR } from '../../../constants/error-messages';
+import { FlutterDatasource } from '.';
```
Comment (viceice) [inline lib/modules/datasource/flutter-version/index.spec.ts:5]:
```suggestion
import { FlutterVersionDatasource } from '.';
```

File: lib/modules/datasource/flutter-version/index.spec.ts [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+const baseUrl = 'https://storage.googleapis.com';
+const urlPath = '/flutter_infra_release/releases/releases_linux.json';
+const datasource = FlutterDatasource.id;
```
Comment (viceice) [inline lib/modules/datasource/flutter-version/index.spec.ts:9]:
```suggestion
const datasource = FlutterVersionDatasource.id;
```

**Review by viceice — COMMENTED** [review]
You can apply all suggestions as batch from github review page 😉

**Comment by zeshuaro** [comment]
Thanks for sharing the batch commit trick, I didn't know we can do that on GitHub.

File: lib/modules/datasource/flutter-version/types.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
+export interface FlutterCurrentRelease {
+  beta: string;
+  dev: string;
+  stable: string;
+}
+
+export interface FlutterRelease {
+  hash: string;
+  channel: string;
```
Comment (viceice) [inline lib/modules/datasource/flutter-version/types.ts:9]:
```suggestion
  channel: 'stable' | 'beta' | 'dev';
```
small improvement for stronger typing 😉

File: lib/modules/datasource/flutter-version/__fixtures__/index.json [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+    {
+      "hash": "184e5871beea4a9b1cbd65dcc57118a7d97285f3",
+      "channel": "dev",
+      "version": "2.5.0-1.0.pre",
+      "release_date": "2021-07-27T21:03:05.871866Z",
+      "archive": "dev/linux/flutter_linux_2.5.0-1.0.pre-dev.tar.xz",
+      "sha256": "7741b4950a414a1e70d3afb1bbdd5e485476e0afd4a399b8f3be93f6bebdae49"
+    }
+  ]
+}
```
Comment (viceice) [inline lib/modules/datasource/flutter-version/__fixtures__/index.json:340]:
```suggestion
}
```

Comment (viceice) [inline lib/modules/datasource/flutter-version/__fixtures__/index.json:340]:
```suggestion
}
```

Comment (viceice) [reply lib/modules/datasource/flutter-version/__fixtures__/index.json:340]: Please add trailing newline

**Review by rarkins — APPROVED** [review]

**Review by rarkins — DISMISSED** [review]

File: lib/modules/datasource/flutter-version/index.ts [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+      registryUrl,
+      releases: [],
+    };
+    try {
+      const resp = (
+        await this.http.getJson<FlutterResponse>(
+          `${registryUrl}/flutter_infra_release/releases/releases_linux.json`
+        )
+      ).body;
+      const stableRegex = /^\d+\.\d+\.\d+$/;
```
Comment (viceice) [inline lib/modules/datasource/flutter-version/index.ts:37]: please move this to module scope and use `util/regex`

File: lib/util/regex.ts [hunk 19]
```diff
@@ -89,3 +89,5 @@ export function configRegexPredicate(
   }
   return null;
 }
+
+export const stableVersionRegex = /^\d+\.\d+\.\d+$/;
```
Comment (viceice) [inline lib/util/regex.ts:93]: no, not moving the regEx to another file. use the `regEx` function from here and move the regEx back to datasource, but outside of the function.

Comment (zeshuaro) [reply lib/util/regex.ts]: Sorry I misunderstood, just updated this. Hopefully I got it right this time 🙂

File: lib/modules/datasource/flutter-version/index.ts [hunk 20]
```diff
@@ -0,0 +1,60 @@
+import { regEx } from '../../../util/regex';
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig, ReleaseResult } from '../types';
+import type { FlutterResponse } from './types';
+
+export const stableVersionRegex = /^\d+\.\d+\.\d+$/;
```
Comment (viceice) [inline lib/modules/datasource/flutter-version/index.ts:6]:
```suggestion
export const stableVersionRegex = regEx(/^\d+\.\d+\.\d+$/);
```
use it here. 🤗

File: lib/modules/datasource/flutter-version/index.ts [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+      const resp = (
+        await this.http.getJson<FlutterResponse>(
+          `${registryUrl}/flutter_infra_release/releases/releases_linux.json`
+        )
+      ).body;
+      result.releases = resp.releases
+        // The API response contains a stable version being released as a non-stable
+        // release. And so we filter out these releases here.
+        .filter(({ version, channel }) => {
+          if (regEx(stableVersionRegex).test(version)) {
```
Comment (viceice) [inline lib/modules/datasource/flutter-version/index.ts:44]:
```suggestion
          if (stableVersionRegex.test(version)) {
```

**Review by viceice — COMMENTED** [review]
nearly done

**Review by viceice — APPROVED** [review]

**Pull request merged**